cron/daily/51-synchronize-paypal-ledger.ts

```diff
@@ -0,0 +1,336 @@
+/**
+ * A script to reconcile PayPal ledgers with the database. Implemented as a CRON job, bu can safely be run manually;
+ * especially on longer period of time (we actually ran it to reconcile the ledger of 2022).
+ */
+
+import '../../server/env';
+
+import { groupBy } from 'lodash';
+import moment from 'moment';
+
+import FEATURE from '../../server/constants/feature';
+import logger from '../../server/lib/logger';
+import { getHostsWithPayPalConnected, listPayPalTransactions } from '../../server/lib/paypal';
+import { reportErrorToSentry } from '../../server/lib/sentry';
+import { parseToBoolean } from '../../server/lib/utils';
+import models, { sequelize } from '../../server/models';
+import { paypalRequestV2 } from '../../server/paymentProviders/paypal/api';
+import { recordPaypalCapture } from '../../server/paymentProviders/paypal/payment';
+import { PaypalCapture, PaypalTransactionSearchResult } from '../../server/types/paypal';
+
+const LIMITED_TO_HOST_SLUGS = process.env.HOST ? process.env.HOST.split(',') : null;
+const START_DATE = process.env.START_DATE ? moment(process.env.START_DATE) : moment().subtract(2, 'day');
+const END_DATE = process.env.END_DATE ? moment(process.env.END_DATE) : moment(START_DATE).add(1, 'day');
+const DRY_RUN = process.env.DRY_RUN ? parseToBoolean(process.env.DRY_RUN) : true;
+
+// Filter out transactions that are not related to contributions
+// See https://developer.paypal.com/docs/transaction-search/transaction-event-codes/
+const WATCHED_EVENT_TYPES = [
+  // Subscription
+  'T0002',
+  // PayPal checkout. These ones are now necessarily recorded in the database, but before we moved to "AUTHORIZE" intent
+  // instead of "CAPTURE" (in https://github.com/opencollective/opencollective-frontend/pull/8601) we got some cases where
+  // the transactions were never recorded. We'll iterate on these in a follow-up PR.
+  // 'T0006',
+];
+
+// Ignore some hosts, usually because they haven't enabled transactions search API yet
+const IGNORED_HOSTS = [
+  // Transactions search not enabled
```

Comment:
Working on enabling them in https://github.com/opencollective/opencollective/issues/6392